thriftrw/protocol/binary.pyx

```diff
@@ -98,6 +99,306 @@ cdef class BinaryProtocolReader(object):
+            self.skip_list()
+
+    cdef void skip_struct(self) except *:
+        header = self.read_field_begin()
+        while header.type != -1:
+            self.skip(header.type)
+            header = self.read_field_begin()
+
+    cdef void skip_map(self) except *:
+        # TODO: This could be optimized to precompute the expected
+        # size for fixed width types
+        cdef MapHeader header = self.read_map_begin()
+        for _ in range(header.size):
+            self.skip(header.ktype)
+            self.skip(header.vtype)
+
+    cdef void skip_list(self) except *:
+        # TODO: This could be optimized to precompute the expected
+        # size for fixed width types
+        cdef ListHeader header = self.read_list_begin()
+        for _ in range(header.size):
+            self.skip(header.type)
+
+    cdef void skip_set(self) except *:
+        # TODO: This could be optimized to precompute the expected
+        # size for fixed width types
+        cdef SetHeader header = self.read_set_begin()
+        for _ in range(header.size):
+            self.skip(header.type)
+
+    cdef void _read(self, char* data, int count) except *:
+        self.reader.read(data, count)
+
+    cdef int8_t _byte(self) except *:
+        cdef char c
+        self._read(&c, 1)
+        return <int8_t>c
+
+    cdef int16_t _i16(self) except *:
+        cdef char data[2]
+        self._read(data, 2)
+        return be16toh((<int16_t*>data)[0])
+
+    cdef int32_t _i32(self) except *:
+        cdef char data[4]
+        self._read(data, 4)
+        return be32toh((<int32_t*>data)[0])
+
+    cdef int64_t _i64(self) except *:
+        cdef char data[8]
+        self._read(data, 8)
+        return be64toh((<int64_t*>data)[0])
+
+    cdef double _double(self) except *:
+        cdef int64_t value = self._i64()
+        return (<double*>(&value))[0]
+
+    cdef bint read_bool(self) except *:
+        """Reads a boolean."""
+        return self._byte() == 1
+
+    cdef int8_t read_byte(self) except *:
+        """Reads a byte."""
+        return self._byte()
+
+    cdef double read_double(self) except *:
+        """Reads a double."""
+        return self._double()
+
+    cdef int16_t read_i16(self) except *:
+        """Reads a 16-bit integer."""
+        return self._i16()
+
+    cdef int32_t read_i32(self) except *:
+        """Reads a 32-bit integer."""
+        return self._i32()
+
+    cdef int64_t read_i64(self) except *:
+        """Reads a 64-bit integer."""
+        return self._i64()
+
+    cdef bytes read_binary(self):
+        """Reads a binary blob."""
+        cdef int32_t length = self._i32()
+        return self.reader.take(length)
+
+    cdef void skip_binary(self) except *:
+        cdef int32_t length = self._i32()
+        self.reader.skip(length)
+
+    cdef FieldHeader read_field_begin(self):
+        cdef int8_t field_type = self._byte()
+        if field_type == STRUCT_END:
+            return FieldHeader(-1, -1)
```

Comment:
make sure to document this contract somewhere obvious like the docs for ProtocolReader.